pkg/scheduler/plugins/capacity/capacity.go

```diff
@@ -220,133 +402,387 @@ func (cp *capacityPlugin) OnSessionOpen(ssn *framework.Session) {
 
 		return 1
 	})
+}
 
-	ssn.AddReclaimableFn(cp.Name(), func(reclaimer *api.TaskInfo, reclaimees []*api.TaskInfo) ([]*api.TaskInfo, int) {
-		var victims []*api.TaskInfo
-		allocations := map[api.QueueID]*api.Resource{}
+func (cp *capacityPlugin) onSessionOpenWithHierarchy(ssn *framework.Session) {
+	// Set the root queue
+	cp.rootQueue = rootQueueID
+	ssn.Queues[api.QueueID(cp.rootQueue)] = &api.QueueInfo{
+		UID:   api.QueueID(cp.rootQueue),
+		Name:  cp.rootQueue,
+		Queue: &scheduling.Queue{},
+	}
 
-		for _, reclaimee := range reclaimees {
-			job := ssn.Jobs[reclaimee.Job]
-			attr := cp.queueOpts[job.Queue]
+	// Initialize queue attributes
+	for _, queue := range ssn.Queues {
+		_, found := cp.queueOpts[queue.UID]
+		if found {
+			continue
+		}
 
-			if _, found := allocations[job.Queue]; !found {
-				allocations[job.Queue] = attr.allocated.Clone()
+		attr := cp.newQueueAttr(queue)
+		cp.queueOpts[queue.UID] = attr
+		err := cp.updateParents(queue, ssn)
+		if err != nil {
+			klog.Errorf("Failed to update Queue <%s> attributes, error: %v", queue.Name, err)
+			return
+		}
+	}
+
+	for _, job := range ssn.Jobs {
+		klog.V(4).Infof("Considering Job <%s/%s>.", job.Namespace, job.Name)
+		attr := cp.queueOpts[job.Queue]
+		if len(attr.children) > 0 {
+			klog.Errorf("The Queue <%s> of Job <%s/%s> is not leaf queue", attr.name, job.Namespace, job.Name)
+			return
```

Comment:
A job being unexpectedly submitted to the parent queue may cause the main process of the current `onSessionOpenWithHierarchy()` to return prematurely. Is the consequence too severe? 

The restriction that "a job cannot be submitted to the parent queue" should perhaps be checked elsewhere, so that the impact of any unexpected job can be contained within the job itself.

Reply:
I have already added this logic in the validate webhook of the job to ensure that jobs can only be submitted to leaf queues.